enhancements/network/host-level-openvswitch.md

```diff
@@ -0,0 +1,138 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenVSwitch (OVS) is a core component to OpenShift networking providers (openshift-sdn, OVN-Kubernetes). This enhancement proposes running OVS as a SystemD service. Currently, it is deployed as a DaemonSet.
+
+However, we must find a way to deploy OVS such that it matches administrators' expectations. That means it should, as much as possible, conform to existing Kubernetes semantics. In other words, the API should be the source of truth, for both desired state and current status.
+
+We propose a "sidecar" DaemonSet that ensures OVS matches a configured state. It also reflects OVS' status back in to the API and exposes logs.
+
+## Motivation
+
+Firstly, running OVS as a SystemD unit matches the technical reality:
+1. Only one ovs-vswitchd process can run on a node. Multitenancy is built in to the daemon, not the kernel.
+2. ovs-vswitchd is tightly coupled, protocol-wise, with the kernel
+3. ovs-vswitchd is in the critical path for networking. Stopping it blocks **all new flows**.
+
+Thus, the proposed solution has the following advantages:
+1. Simplifies the burden on OVS maintainers by shrinking the test matrix of kernel-daemon versions.
+2. Removes a known source of packet loss during upgrades
+3. Allows OVS to be shared between multiple networking components (openshift-sdn, CNV).
+
+### Goals & Requirements
+
+1. OVS is run in a SystemD unit provided by the base OS, but is still manageable as if it were a Kubernetes Pod:
+  - kubectl logs works
+  - Liveness- and Readiness- probes reflect the unit’s state
+  - Note: it is a non-goal that deleting the sidecar Pod stops OVS
+2. OVS is stable and more available than the current state.
+
+### Non-Goals
+
+1. Provide a generic solution for managing SystemD units via Kubernetes
+2. Reconcile all possible configuration changes (e.g. stopping / disabling OVS).
+
+## Proposal
+
+There is a pod that acts as a “sidecar” or “babysitter” for the openvswitch systemd unit. It needs to
+
+- Ensure the OVS unit is enabled and started
+- Provide a Liveness probe
+- Provide a Readiness probe
+- Apply any desired OVS configuration (loglevel, memory usage, etc)
+- Print log messages to stdout
+
+Configuration for OVS is always presented as a set of key-value pairs that are inserted in to a configuration database. If needed, we should allow custom configuration to be expressed as a ConfigMap, which is reconciled to the database.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+As a prototype, this can be implemented with just a few shell scripts. Ideally, it should be ultimately implemented as a go program that talks to the SystemD DBus API directly.
```

Comment:
I belive the code will be hosted in it's own repo? so that every CNI that depends on OVS can fetch it, rather than keeping their own copy of scripts/code to run in sidecar?

Reply:
I am just wondering if we will hit a scenario, where we use multiple network interface and both the CNI relies on OVS, will this implementation run two sidecar (per CNI) to manage the host OVS state?

Reply:
Good point. As a first solution, we should add a flag in the CNO's configuration for people to request openvswitch enablement.